packages/flutter/lib/src/services/predictive_back_event.dart

```diff
@@ -0,0 +1,106 @@
+
+/// Enum representing the edge from which a swipe starts in a back gesture.
+///
+/// This is used in [PredictiveBackEvent] to indicate the starting edge of the
+/// swipe gesture.
+enum SwipeEdge {
+  /// Indicates that the swipe gesture starts from the left edge of the screen.
+  left,
+
+  /// Indicates that the swipe gesture starts from the right edge of the screen.
+  right,
+}
+
+/// Object used to report back gesture progress in Android.
+///
+/// Holds information about the touch event, swipe direction and the animation
+/// progress that predictive back animations should seek to.
+@immutable
+class PredictiveBackEvent {
+  /// Creates a new [PredictiveBackEvent] instance.
+  const PredictiveBackEvent._({
+    required this.x,
+    required this.y,
+    required this.progress,
+    required this.swipeEdge,
+  }) : assert(progress >= 0.0 && progress <= 1.0);
+
+  /// Creates an [PredictiveBackEvent] from a Map, typically used when converting
+  /// data received from a platform channel.
+  factory PredictiveBackEvent.fromJSON(Map<String, dynamic> json) {
+    return PredictiveBackEvent._(
+      x: (json['x'] as num?)?.toDouble(),
+      y: (json['y'] as num?)?.toDouble(),
+      progress: (json['progress'] as num).toDouble(),
+      swipeEdge: SwipeEdge.values[json['swipeEdge'] as int],
+    );
+  }
+
+  /// The global X location of the touch point, or `null` if the event is from a
+  /// button press.
+  final double? x;
+
+  /// The global Y location of the touch point, or `null` if the event is from a
+  /// button press.
+  final double? y;
+
+  /// Returns a value between 0 and 1 representing how far along the back
+  /// gesture is.
+  ///
+  /// This value is driven by the horizontal location of the touch point, and
+  /// should be used as the fraction to seek the predictive back animation with.
+  /// Specifically,
+  ///
+  /// - The progress is 0 when the touch is at the starting edge of the screen
+  ///   (left or right), and the animation should seek to its start state.
+  /// - The progress is approximately 1 when the touch is at the opposite side
+  ///   of the screen, and the animation should seek to its end state. Exact end
+  ///   value may vary depending on screen size.
+  ///
+  /// When the gesture is canceled, the progress value continues to update,
+  /// animating back to 0 until the cancellation animation completes.
+  ///
+  /// In-between locations are linearly interpolated based on horizontal
+  /// distance from the starting edge and smooth clamped to 1 when the distance
+  /// exceeds a system-wide threshold.
+  final double progress;
+
+  /// The screen edge from which the swipe gesture starts.
+  final SwipeEdge swipeEdge;
+
+  /// Determines if the event was triggered by a button press.
+  ///
+  /// It returns true when the back button is pressed, such as when the user
+  /// opts for 3-button navigation. In cases of gesture navigation, it returns
+  /// false.
+  bool get isButtonEvent =>
```

Comment:
(Note that `x` and `y` have since changed to `touchOffset`)

@maRci002 Did you include this in `isButtonEvent` for a reason? It seems like it could be possible that `progress == 0.0 && touchOffset == Offset.zero` for the start of a back gesture if it's not stated somewhere that it can't?

@LongCatIsLooong Yeah the 3 button navigation is just referring to a system back button. Especially older Android phones had these kinds of buttons along the bottom of the screen. I'll clarify in the docs a bit.

Reply:
Tapping the system back button is not considered predictive back, and it is handled by Flutter by directly popping.

Reply:
thanks for the explanation. I looked at the Android predictive back gesture codelab, to enable predictive back you have to set the `android:enableOnBackInvokedCallback` flag in AndroidManifest.xml. 

Does setting the flag also disable the old callpath that we use to handle directly popping (so `WidgetBindingObserver.didPopRoute` et.al will never be called if the flag is set), or both the old and the new callbacks will be called when the system back button is pressed?

Reply:
The former. When the flag is not set, a back button comes in through `popRoute`, otherwise it comes via `commitBackGesture`.

And to follow up on the question about `x == 0` etc., I tried logging the output of [getTouchX](https://developer.android.com/reference/android/window/BackEvent#getTouchX()) directly in the engine when I press the system back button, and indeed it is giving me `0.0` despite saying in those docs that it should give `NaN`. I guess that means that we should keep the logic as-is. I'll leave a comment about this.